FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +30,95 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		let request = coreDataFeedImageFetchRequest()
+		queue.async {
+			do {
+				guard let fetchedFeed = try context.fetch(request) as? [CoreDataFeedImage], let timestamp = fetchedFeed.first?.timestamp else {
+					completion(.empty)
+					return
+				}
+
+				let feed = fetchedFeed.toSortedLocal()
+				completion(.found(feed: feed, timestamp: timestamp))
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		queue.async(flags: .barrier) { [weak self] in
```

Comment:
There's no need to weakify `self` in this context.

Moreover, instead of using a custom background queue, you should use the context's queue with `context.perform`:

```suggestion
		context.perform {
```

Reply:
fixed